src/spaces/deligne.jl

```diff
@@ -12,52 +12,68 @@ The Deligne tensor product also works in the type domain and for sectors and ten
+function ⊠(V₁::VectorSpace, V₂::VectorSpace)
+    check_fields_deligne(V₁, V₂)
+    return (V₁ ⊠ one(V₂)) ⊗ (one(V₁) ⊠ V₂)
+end
 
-# define deligne products with empty tensor product: just add a trivial sector of the type of the empty space to each of the sectors in the non-empty space
-function ⊠(V::GradedSpace, P₀::ProductSpace{<:ElementarySpace{ℂ},0})
+# define deligne products with empty tensor product: just add a trivial sector of the type
+# of the empty space to each of the sectors in the non-empty space
+function ⊠(V::GradedSpace, P₀::ProductSpace{<:ElementarySpace,0})
+    check_fields_deligne(V, P₀)
     I₁ = sectortype(V)
     I₂ = sectortype(P₀)
     return Vect[I₁ ⊠ I₂](ifelse(isdual(V), dual(c), c) ⊠ one(I₂) => dim(V, c)
                          for c in sectors(V); dual=isdual(V))
 end
 
-function ⊠(P₀::ProductSpace{<:ElementarySpace{ℂ},0}, V::GradedSpace)
+function ⊠(P₀::ProductSpace{<:ElementarySpace,0}, V::GradedSpace)
+    check_fields_deligne(V, P₀)
     I₁ = sectortype(P₀)
     I₂ = sectortype(V)
     return Vect[I₁ ⊠ I₂](one(I₁) ⊠ ifelse(isdual(V), dual(c), c) => dim(V, c)
                          for c in sectors(V); dual=isdual(V))
 end
 
-function ⊠(V::ComplexSpace, P₀::ProductSpace{<:ElementarySpace{ℂ},0})
+function ⊠(V::ComplexSpace, P₀::ProductSpace{<:ElementarySpace,0})
+    check_fields_deligne(V, P₀)
     I₂ = sectortype(P₀)
     return Vect[I₂](one(I₂) => dim(V); dual=isdual(V))
 end
 
-function ⊠(P₀::ProductSpace{<:ElementarySpace{ℂ},0}, V::ComplexSpace)
+function ⊠(P₀::ProductSpace{<:ElementarySpace,0}, V::ComplexSpace)
+    check_fields_deligne(V, P₀)
     I₁ = sectortype(P₀)
     return Vect[I₁](one(I₁) => dim(V); dual=isdual(V))
 end
 
-function ⊠(P::ProductSpace{<:ElementarySpace{ℂ},0},
-           P₀::ProductSpace{<:ElementarySpace{ℂ},0})
+function ⊠(P::ProductSpace{<:ElementarySpace,0},
+           P₀::ProductSpace{<:ElementarySpace,0})
+    check_fields_deligne(P, P₀)
     I₁ = sectortype(P)
     I₂ = sectortype(P₀)
     return one(Vect[I₁ ⊠ I₂])
 end
 
-function ⊠(P::ProductSpace{<:ElementarySpace{ℂ}}, P₀::ProductSpace{<:ElementarySpace{ℂ},0})
+function ⊠(P::ProductSpace{<:ElementarySpace}, P₀::ProductSpace{<:ElementarySpace,0})
+    check_fields_deligne(P, P₀)
     I₁ = sectortype(P)
     I₂ = sectortype(P₀)
     S = Vect[I₁ ⊠ I₂]
     N = length(P)
     return ProductSpace{S,N}(map(V -> V ⊠ P₀, tuple(P...)))
 end
 
-function ⊠(P₀::ProductSpace{<:ElementarySpace{ℂ},0}, P::ProductSpace{<:ElementarySpace{ℂ}})
+function ⊠(P₀::ProductSpace{<:ElementarySpace,0}, P::ProductSpace{<:ElementarySpace})
+    check_fields_deligne(P, P₀)
     I₁ = sectortype(P₀)
     I₂ = sectortype(P)
     S = Vect[I₁ ⊠ I₂]
     N = length(P)
     return ProductSpace{S,N}(map(V -> P₀ ⊠ V, tuple(P...)))
 end
+
+@noinline function check_fields_deligne(P₁, P₂)
+    return field(P₁) == field(P₂) ||
+           throw(ArgumentError("Deligne products require spaces over the same field"))
+end
```

Comment:
I think I was too quick with this suggestion. The point is that you want to have that, if the compiler can infer that the check is true, that you don't create a string (as this allocates memory). This can be obtained by having the check, and then having the creation of the error message in a separate function that has a `@noinline` decorator. As the body of this function is then not executed if the check is inferably true, no string is being allocated. Check some of the `@noinline` use cases in Base: https://github.com/search?q=repo%3AJuliaLang%2Fjulia+%40noinline&type=code

The current implementation (which I suggested) is not good. It has the check together with the error creation, so when this function is executed, the string is always allocated, even if the check is true. The `@noinline` in this case is also not usefu; it just means that the check itself is also never inlined, and thus there is always a small function call overhead for evaluating the check.

Reply:
Ah, I missed this. I think this should then fix it, I'll also do this for the inner product style checks.